File: sdks/python/a2ui_toolkit/ag_ui_a2ui_toolkit/validate.py
"""Semantic validation of A2UI v0.9 component trees (OSS-162).

Python port of ``a2ui-toolkit/src/validate.ts`` — kept behaviorally identical so
the framework adapters and the middleware agree on what "valid" means. Adds the
semantic checks (catalog membership, required props, child refs, binding
resolution) whose failures otherwise blow up at render time, turning them into
machine-readable errors the recovery loop can feed back to the sub-agent.
"""

from __future__ import annotations

from typing import Any, Optional

# A validation error is a plain dict: {"code", "path", "message"} — JSON-friendly
# so it can ride straight into a prompt / event payload.
A2UIValidationError = dict[str, str]
ValidateA2UIResult = dict[str, Any]  # {"valid": bool, "errors": list[A2UIValidationError]}


def _is_object(v: Any) -> bool:
    return isinstance(v, dict)


def _absolute_path_resolves(path: str, data: Any) -> bool:
    segments = [s for s in path.split("/") if s]
    cursor: Any = data
    for seg in segments:
        if cursor is None or not isinstance(cursor, (dict, list)):
            return False
        if isinstance(cursor, list):
            try:
                idx = int(seg)
            except ValueError:
                return False
            if idx < 0 or idx >= len(cursor):
                return False
            cursor = cursor[idx]
        else:
            if seg not in cursor:
                return False
            cursor = cursor[seg]
    return True


def _collect_child_refs(children: Any) -> list[str]:
    refs: list[str] = []

    def push(v: Any) -> None:
        if isinstance(v, str):
            refs.append(v)
        elif _is_object(v) and isinstance(v.get("componentId"), str):
            refs.append(v["componentId"])

    if isinstance(children, list):
        for v in children:
            push(v)
    elif _is_object(children):
        push(children)
    return refs


def _collect_absolute_binding_paths(node: Any, acc: list[str]) -> list[str]:
    if isinstance(node, list):
        for v in node:
            _collect_absolute_binding_paths(v, acc)
    elif _is_object(node):
        p = node.get("path")
        if isinstance(p, str) and p.startswith("/"):
            acc.append(p)
        for k, v in node.items():
            if k == "path":
                continue
            _collect_absolute_binding_paths(v, acc)
    return acc


def validate_a2ui_components(
    *,
    components: Any,
    data: Optional[dict[str, Any]] = None,
    catalog: Optional[dict[str, Any]] = None,
    validate_bindings: bool = True,
) -> ValidateA2UIResult:
    """Validate a flat A2UI v0.9 component array.

    Structural checks always run. Catalog membership + required-prop checks run
    only when ``catalog`` is supplied. Absolute binding paths (``/foo``) resolve
    against ``data``; relative template paths (``name``) are left alone — they
    resolve per-item inside a repeated template and flagging them would produce
    false positives (and spurious retries).
    """
    errors: list[A2UIValidationError] = []

    # Fail loud on a non-list / empty payload.
    if not isinstance(components, list) or len(components) == 0:
        return {
            "valid": False,
            "errors": [{"code": "empty_components", "path": "components", "message": "A2UI components must be a non-empty array"}],
        }

    ids: set[str] = set()
    seen: set[str] = set()
    for comp in components:
        cid = comp.get("id") if _is_object(comp) else None
        if isinstance(cid, str):
            if cid in seen:
                errors.append({"code": "duplicate_id", "path": f"components[id={cid}]", "message": f"Duplicate component id '{cid}'"})
            seen.add(cid)
            ids.add(cid)

    catalog_components = (catalog or {}).get("components", {}) if catalog else {}

    for i, comp in enumerate(components):
        cid = comp.get("id") if _is_object(comp) else None
        ctype = comp.get("component") if _is_object(comp) else None

        if not isinstance(cid, str) or len(cid) == 0:
            errors.append({"code": "missing_id", "path": f"components[{i}].id", "message": f"Component at index {i} is missing a string 'id'"})
        if not isinstance(ctype, str) or len(ctype) == 0:
            errors.append({"code": "missing_component_type", "path": f"components[{i}].component", "message": f"Component at index {i} is missing a string 'component' type"})

        if catalog and isinstance(ctype, str):
            schema = catalog_components.get(ctype)
            if schema is None:
                errors.append({"code": "unknown_component", "path": f"components[{i}].component", "message": f"Component type '{ctype}' is not in the catalog"})
            else:
                for req in schema.get("required", []) or []:
                    if not _is_object(comp) or req not in comp:
                        errors.append({"code": "missing_required_prop", "path": f"components[{i}].{req}", "message": f"Component '{ctype}' (index {i}) is missing required prop '{req}'"})

        if _is_object(comp):
            for ref in _collect_child_refs(comp.get("children")):
                if ref not in ids:
                    errors.append({"code": "unresolved_child", "path": f"components[{i}].children", "message": f"Child reference '{ref}' does not match any component id"})
            for p in (_collect_absolute_binding_paths(comp, []) if validate_bindings else []):
                if not _absolute_path_resolves(p, data or {}):
                    errors.append({"code": "unresolved_binding", "path": f"components[{i}]", "message": f"Binding path '{p}' does not resolve in the data model"})

    if not any(_is_object(c) and c.get("id") == "root" for c in components):
        errors.append({"code": "no_root", "path": "components", "message": "No component has id 'root'"})

    return {"valid": len(errors) == 0, "errors": errors}

File: sdks/typescript/packages/a2ui-toolkit/src/validate.ts
/**
 * Semantic validation of A2UI v0.9 component trees (OSS-162).
 *
 * The middleware's streaming path only checks *structural* completeness (array
 * closed, each item has a `component` string). This module adds the *semantic*
 * checks whose failures otherwise blow up at render time in `@a2ui/web_core`
 * ("Component not found", "Catalog not found", unresolved bindings) — turning
 * them into machine-readable errors the recovery loop can feed back to the
 * sub-agent.
 *
 * Used by BOTH the adapter (to decide whether to retry) and the middleware (to
 * decide whether to paint) so the two never disagree on what "valid" means.
 */

/** A single, machine-readable validation failure. */
export interface A2UIValidationError {
  code:
    | "empty_components"
    | "missing_id"
    | "missing_component_type"
    | "duplicate_id"
    | "no_root"
    | "unknown_component"
    | "missing_required_prop"
    | "unresolved_child"
    | "unresolved_binding";
  /** A JSON-pointer-ish locator, e.g. `components[2].component`. */
  path: string;
  /** Human/LLM-readable description (fed back to the sub-agent on retry). */
  message: string;
}

export interface ValidateA2UIResult {
  valid: boolean;
  errors: A2UIValidationError[];
}

/**
 * Inline JSON-Schema catalog (mirrors the middleware's `A2UIInlineCatalogSchema`):
 * component name → JSON Schema whose `required` lists mandatory props.
 */
export interface A2UIValidationCatalog {
  components: Record<string, { required?: string[]; properties?: Record<string, unknown>; [k: string]: unknown }>;
}

export interface ValidateA2UIInput {
  components: Array<Record<string, unknown>>;
  /** The surface's data model; used to resolve absolute binding paths. */
  data?: Record<string, unknown>;
  /** When omitted, catalog-dependent checks (membership, required props) are skipped. */
  catalog?: A2UIValidationCatalog;
  /**
   * Resolve absolute binding paths against `data`. Default `true`. Set `false`
   * at the streaming component-close boundary, where the component tree has
   * closed but the data model has not streamed yet — resolving bindings there
   * would false-positive (and trigger spurious retries). The adapter re-runs
   * full validation (bindings included) once the complete args arrive.
   */
  validateBindings?: boolean;
}

/** Does `path` (absolute, e.g. `/items/0/name`) resolve in `data`? */
function absolutePathResolves(path: string, data: unknown): boolean {
  const segments = path.split("/").filter((s) => s.length > 0);
  let cursor: unknown = data;
  for (const seg of segments) {
    if (cursor == null || typeof cursor !== "object") return false;
    if (Array.isArray(cursor)) {
      const idx = Number(seg);
      if (!Number.isInteger(idx) || idx < 0 || idx >= cursor.length) return false;
      cursor = cursor[idx];
    } else {
      if (!(seg in (cursor as Record<string, unknown>))) return false;
      cursor = (cursor as Record<string, unknown>)[seg];
    }
  }
  return true;
}

/** True for a plain (non-array) object. */
function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/**
 * Validate a flat A2UI v0.9 component array.
 *
 * Structural checks always run. Catalog membership + required-prop checks run
 * only when `catalog` is supplied. Absolute binding paths (`/foo`) are resolved
 * against `data`; relative template paths (`name`) are left alone — they resolve
 * per-item inside a repeated template and flagging them would produce false
 * positives (and spurious retries).
 */
export function validateA2UIComponents(input: ValidateA2UIInput): ValidateA2UIResult {
  const { components, data, catalog } = input;
  const validateBindings = input.validateBindings ?? true;
  const errors: A2UIValidationError[] = [];

  // Fail loud on a non-array / empty payload — there is nothing to render and
  // nothing meaningful to feed back, so the caller must not treat it as a
  // recoverable surface silently.
  if (!Array.isArray(components) || components.length === 0) {
    return {
      valid: false,
      errors: [{ code: "empty_components", path: "components", message: "A2UI components must be a non-empty array" }],
    };
  }

  const ids = new Set<string>();
  const seen = new Set<string>();
  for (const comp of components) {
    const id = isObject(comp) ? comp.id : undefined;
    if (typeof id === "string") {
      if (seen.has(id)) {
        errors.push({ code: "duplicate_id", path: `components[id=${id}]`, message: `Duplicate component id '${id}'` });
      }
      seen.add(id);
      ids.add(id);
    }
  }

  components.forEach((comp, i) => {
    const id = isObject(comp) ? comp.id : undefined;
    const type = isObject(comp) ? comp.component : undefined;

    if (typeof id !== "string" || id.length === 0) {
      errors.push({ code: "missing_id", path: `components[${i}].id`, message: `Component at index ${i} is missing a string 'id'` });
    }
    if (typeof type !== "string" || type.length === 0) {
      errors.push({
        code: "missing_component_type",
        path: `components[${i}].component`,
        message: `Component at index ${i} is missing a string 'component' type`,
      });
    }

    // Catalog membership + required props (only when a catalog is supplied).
    if (catalog && typeof type === "string") {
      const schema = catalog.components[type];
      if (!schema) {
        errors.push({
          code: "unknown_component",
          path: `components[${i}].component`,
          message: `Component type '${type}' is not in the catalog`,
        });
      } else {
        for (const req of schema.required ?? []) {
          if (!isObject(comp) || !(req in comp)) {
            errors.push({
              code: "missing_required_prop",
              path: `components[${i}].${req}`,
              message: `Component '${type}' (index ${i}) is missing required prop '${req}'`,
            });
          }
        }
      }
    }

    // Child references must resolve to existing component ids.
    if (isObject(comp)) {
      collectChildRefs(comp.children).forEach((ref) => {
        if (!ids.has(ref)) {
          errors.push({
            code: "unresolved_child",
            path: `components[${i}].children`,
            message: `Child reference '${ref}' does not match any component id`,
          });
        }
      });

      // Absolute binding paths must resolve against the data model (unless
      // deferred — see `validateBindings`).
      if (validateBindings) collectAbsoluteBindingPaths(comp).forEach((p) => {
        if (!absolutePathResolves(p, data ?? {})) {
          errors.push({
            code: "unresolved_binding",
            path: `components[${i}]`,
            message: `Binding path '${p}' does not resolve in the data model`,
          });
        }
      });
    }
  });

  if (!components.some((c) => isObject(c) && c.id === "root")) {
    errors.push({ code: "no_root", path: "components", message: "No component has id 'root'" });
  }

  return { valid: errors.length === 0, errors };
}

/** Pull child-id references out of a `children` value (array of ids or {componentId,...}). */
function collectChildRefs(children: unknown): string[] {
  const refs: string[] = [];
  const push = (v: unknown) => {
    if (typeof v === "string") refs.push(v);
    else if (isObject(v) && typeof v.componentId === "string") refs.push(v.componentId);
  };
  if (Array.isArray(children)) children.forEach(push);
  else if (isObject(children)) push(children);
  return refs;
}

/** Recursively collect absolute (`/…`) binding paths from a component's props. */
function collectAbsoluteBindingPaths(node: unknown, acc: string[] = []): string[] {
  if (Array.isArray(node)) {
    node.forEach((v) => collectAbsoluteBindingPaths(v, acc));
  } else if (isObject(node)) {
    if (typeof node.path === "string" && node.path.startsWith("/")) acc.push(node.path);
    for (const [k, v] of Object.entries(node)) {
      if (k === "path") continue;
      collectAbsoluteBindingPaths(v, acc);
    }
  }
  return acc;
}
